### PR TITLE
Fix Issue 19429 - indexOf("a", "b") should be nothrow/@nogc

### PR DESCRIPTION
**Pull request by CyberShadow**
Please see the individual commits' commit messages for details.

### CODE REVIEW

**Comment by CyberShadow** [comment]
This is a non-trivial change, let's wait at least 24 hours for reviews.

File: std/string.d [hunk 1]
```diff
@@ -871,20 +871,44 @@ private template _indexOfStr(CaseSensitive cs)
...
+        in size_t startIdx)
+@safe
+if (isSomeChar!Char1 && isSomeChar!Char2)
+{
+    if (startIdx < s.length)
+    {
+        ptrdiff_t foundIdx = indexOf(s[startIdx .. $], sub);
+        if (foundIdx != -1)
+        {
+            return foundIdx + cast(ptrdiff_t) startIdx;
```
Comment (UplinkCoder) [inline std/string.d]: you cloud do foundIdx += cast(ptrdiff_t) startIdx;

and then return foundIdx at the end, which would safe one ret instruction and some branch predictor pain,

Comment (CyberShadow) [reply std/string.d]: Also, this is not my code. The function was duplicated in order to allow it to have its own inferred attributes, due to case sensitivity no longer being variable at runtime.

Comment (UplinkCoder) [reply std/string.d]: Hmm dmd amazes es always :)

Comment (UplinkCoder) [reply std/string.d]: my suggestion was different though
```
    ptrdiff_t foundIdx = -1;
    if (startIdx < s.length)
    {
        foundIdx = indexOf(s[startIdx .. $], sub);
        if (foundIdx != -1)
        {
            foundIdx += cast(ptrdiff_t) startIdx;
        }    
    }
    return foundIdx;
```

Comment (CyberShadow) [reply std/string.d]: Tried that too, still more instructions...

**Comment by CyberShadow** [comment]
BuildKite failure looks unrelated.

**Pull request merged**